File: app/src/main/java/com/djavid/bitcoinrate/util/PresenterProvider.java
package com.djavid.bitcoinrate.util;

import com.djavid.bitcoinrate.core.Presenter;
import com.djavid.bitcoinrate.presenter.implementations.RateFragmentPresenterImpl;
import com.djavid.bitcoinrate.presenter.implementations.TickerFragmentPresenterImpl;

import java.util.HashMap;
import java.util.Map;


public class PresenterProvider {

    private Map<String, Presenter> presenterMap;

    public PresenterProvider() {
        presenterMap = new HashMap<>();
    }

    public <T extends Presenter> T getPresenter(String presenterId, Class<T> c) {
        createPresenter(presenterId);
        return c.cast(presenterMap.get(presenterId));
    }

    private void createPresenter(String presenterId) {
        if (presenterMap.containsKey(presenterId)) return;

        switch (presenterId){
            case "rate_fragment":
                presenterMap.put(presenterId, new RateFragmentPresenterImpl());
                break;
            case "ticker_fragment":
                presenterMap.put(presenterId, new TickerFragmentPresenterImpl());
                break;
        }
    }

    public void removePresenter(String presenterId) {
        if (presenterMap.containsKey(presenterId))
            presenterMap.remove(presenterId);
    }

}

File: app/src/main/java/com/djavid/bitcoinrate/util/PresenterProvider.kt
package com.djavid.bitcoinrate.util

import com.djavid.bitcoinrate.core.Presenter
import com.djavid.bitcoinrate.presenter.implementations.RateFragmentPresenterImpl
import com.djavid.bitcoinrate.presenter.implementations.TickerFragmentPresenterImpl
import java.util.*


class PresenterProvider {

    private val presenterMap: MutableMap<String, Presenter<*, *, *>>

    init {
        presenterMap = HashMap<String, Presenter>()
    }

    fun <T : Presenter<*, *, *>> getPresenter(presenterId: String, c: Class<T>): T {
        createPresenter(presenterId)
        return c.cast(presenterMap[presenterId])
    }

    private fun createPresenter(presenterId: String) {
        if (presenterMap.containsKey(presenterId)) return

        when (presenterId) {
            "rate_fragment" -> presenterMap[presenterId] = RateFragmentPresenterImpl()
            "ticker_fragment" -> presenterMap[presenterId] = TickerFragmentPresenterImpl()
        }
    }

    fun removePresenter(presenterId: String) {
        if (presenterMap.containsKey(presenterId))
            presenterMap.remove(presenterId)
    }

}
